processing/src/main/java/io/druid/query/monomorphicprocessing/SpecializationService.java

```diff
@@ -0,0 +1,391 @@
+ * no-arg constructor and must be stateless (no fields).
+ *
+ * @see SpecializationState
+ */
+public final class SpecializationService
+{
+  private static final Logger LOG = new Logger(SpecializationService.class);
+
+  private static final Unsafe UNSAFE;
+  static {
+    try {
+      Field theUnsafe = Unsafe.class.getDeclaredField("theUnsafe");
+      theUnsafe.setAccessible(true);
+      UNSAFE = (Unsafe) theUnsafe.get(null);
+    }
+    catch (Exception e) {
+      throw new RuntimeException("Cannot access Unsafe methods", e);
+    }
+  }
+
+  /**
+   * If true, specialization is not actually done, an instance of prototypeClass is used as a "specialized" instance.
+   * Useful for analysis of generated assembly with JITWatch (https://github.com/AdoptOpenJDK/jitwatch), because
+   * JITWatch shows only classes present in the loaded JAR (prototypeClass should be), not classes generated during
+   * runtime.
+   */
+  private static final boolean fakeSpecialize = Boolean.getBoolean("fakeSpecialize");
+
+  /**
+   * Number of loop iterations, accounted via {@link SpecializationState#accountLoopIterations(long)} in
+   * {@link WindowedLoopIterationCounter} during the last hour window, after which WindowedLoopIterationCounter decides
+   * to specialize class for the specific runtimeShape. The default value is chosen to be so that the specialized
+   * class will likely be compiled with C2 HotSpot compiler with the default values of *BackEdgeThreshold options.
+   */
+  private static final long triggerSpecializationIterationsThreshold =
+      Integer.getInteger("triggerSpecializationIterationsThreshold", 10_000);
+
+  private static final ExecutorService classSpecializationExecutor = Execs.singleThreaded("class-specialization-%d");
+
+  private static final AtomicLong specializedClassCounter = new AtomicLong();
+
+  private static final ClassValue<PerPrototypeClassState> perPrototypeClassState =
+      new ClassValue<PerPrototypeClassState>()
+      {
+        @Override
+        protected PerPrototypeClassState computeValue(Class<?> type)
+        {
+          return PerPrototypeClassState.of(type);
+        }
+      };
+
+  /**
+   * @param <T> type of query processing algorithm
+   * @see SpecializationService class-level javadoc for details
+   */
+  public static <T> SpecializationState<T> getSpecializationState(
+      Class<? extends T> prototypeClass,
+      String runtimeShape
+  )
+  {
+    return getSpecializationState(prototypeClass, runtimeShape, ImmutableMap.<Class<?>, Class<?>>of());
+  }
+
+  /**
+   * @param classRemapping classes, that should be replaced in the bytecode of the given prototypeClass when specialized
+   * @see #getSpecializationState(Class, String)
+   */
+  @SuppressWarnings("unchecked")
+  public static <T> SpecializationState<T> getSpecializationState(
+      Class<? extends T> prototypeClass,
+      String runtimeShape,
+      ImmutableMap<Class<?>, Class<?>> classRemapping
```

Comment:
it appears that classRemapping is always passed empty map, why do we need it as argument?